Select component

lukas-hetzenecker opened the pull request:
This is an initial pull request for a new component called option.
It allows the user to select one option ouf of a group of possible options.

It will be used by the HDMI-CEC platform.

lukas-hetzenecker committed:
Initial support for option component

lukas-hetzenecker committed:
Added demo for option component

lukas-hetzenecker committed:
Added command-line option

lukas-hetzenecker commented:
The different options will be represented with buttons in the UI:

![screenshot from 2015-12-27 02-53-11](https://cloud.githubusercontent.com/assets/2315617/12008577/1460cac4-ac45-11e5-9067-2c515b10aa81.png)
![screenshot from 2015-12-27 02-53-35](https://cloud.githubusercontent.com/assets/2315617/12008578/146294f8-ac45-11e5-862c-8b076d201091.png)

lukas-hetzenecker commented:
The correspondig UI changes are in balloob/home-assistant-polymer#16

balloob commented:
Dang, I wished you would have reached out to me before stating coding. I fully agree with the idea but I envisioned it slightly differently:

Create a new variables component that can hold different types of variables. One being a boolean (represented in the UI as a switch) and an option type, like yours. Instead of leveraging the more info screen to allow users to pick options, use a [paper-dropdown-menu](https://elements.polymer-project.org/elements/paper-dropdown-menu?view=demo:demo/index.html&active=paper-dropdown-menu) inside the overview.

I think the command options platform is conflating concerns of executing something and allowing the user to pick an option. Instead, this should be done using automation where a different command is called based on the value of a variable.

lukas-hetzenecker commented:
Thanks for your feedback. I completely agree about the paper-dropdown-menu component, this is obviously the better option.

But I don't quite understand why there should be only one component that handles different types of variables - in my view the intended meaning of the component is different: whereas a boolean represents a on/off state, the option component is used for selecting one state out of many. I honestly see no simplification of the code by combining them.

I also have a use case where the state of the component itself is not know and only an action can be triggered (in my case that is emitting IR signals). That's simple to solve using paper-buttons, but I don't know how this can be done using dropdown-menus. Do you think just adding a non-selectable "(unspecified)" option is enough for this?

Please tell me how I should proceed with this pull request.

balloob commented:
Gave it some more thought. Let's keep it as a stand alone component but have it be one without any platforms. Just one that can be configured via the configuration and that's that.

Let's update the name to something less ambiguous, what about `input_select`? 

See the [scene component](https://github.com/balloob/home-assistant/blob/dev/homeassistant/components/scene.py#L50-L53) for an example of how to use the EntityComponent without platforms.

The options would be configured via `configuration.yaml`. The config could look like this:

```yaml
input_select:
  name: Channel chooser
  allow_no_choice: yes
  options:
    - Channel 1
    - Channel 2
    - Channel 3
```

balloob commented:
CI also reported some style violations: https://travis-ci.org/balloob/home-assistant/jobs/98928341#L828-L836

balloob commented:
Could you also include a simple test to ensure that the component works as expected?

balloob commented on homeassistant/components/option/__init__.py:
```diff
@@ -0,0 +1,121 @@
+    descriptions = load_yaml_config_file(
+        os.path.join(os.path.dirname(__file__), 'services.yaml'))
+
+    print("descriptions", descriptions)
```
debug statement.

balloob commented on homeassistant/components/option/__init__.py:
```diff
@@ -0,0 +1,121 @@
+    return True
+
+
+class OptionDevice(Entity):
```
Add a `should_poll` property that returns `False` so the entity component will not poll the entities.

balloob commented on homeassistant/components/option/__init__.py:
```diff
@@ -0,0 +1,121 @@
+    def state_attributes(self):
+        """ Returns optional state attributes. """
+        data = {
+            ATTR_OPTION: self.option,
```
No need to capture the current option in the state attributes, it is already represented as the state.

balloob commented on homeassistant/components/option/services.yaml:
```diff
@@ -0,0 +1,13 @@
+
+  fields:
+    entity_id:
+      description: Name(s) of entities to turn on
```
This description is wrong.

balloob commented on homeassistant/const.py:
```diff
@@ -147,6 +147,8 @@
 SERVICE_MOVE_DOWN = 'move_down'
 SERVICE_STOP = 'stop'
 
+SERVICE_SET_OPTION = 'set_option'
```
No need to put it in the `const` file, just keep it inside the component as this will not be reused by other components.

lukas-hetzenecker committed:
Rename component to 'select'.
The aggregation of options is named 'select', the entity that handles it 'SelectableDevice'.
Each individual element in this selectable device is named 'option'. An 'option' can be selected.
This follows the semantics of the HTML select / option tags.

lukas-hetzenecker committed:
remove debugging statement

lukas-hetzenecker committed:
move global SERVICE_SET_OPTION to component

balloob commented on homeassistant/components/select/__init__.py:
```diff
@@ -0,0 +1,110 @@
+        """ Returns the list of available options for this entity. """
+        return []
+
+    def select(self, option, **kwargs):
```
No need to add `kwargs` here.

balloob commented on homeassistant/components/select/__init__.py:
```diff
@@ -0,0 +1,110 @@
+
+def setup(hass, config):
+    """ Setup selects. """
+    component = EntityComponent(_LOGGER, DOMAIN, hass,
```
No need to define `SCAN_INTERVAL` or `DISCOVERY_PLATFORMS` since you're not using it.

balloob commented on homeassistant/components/select/__init__.py:
```diff
@@ -0,0 +1,110 @@
+"""
+homeassistant.components.select
```
Can you name it `input_select`? `select` is too ambiguous.

balloob commented on homeassistant/components/select/__init__.py:
```diff
@@ -0,0 +1,110 @@
+ENTITY_ID_FORMAT = DOMAIN + '.{}'
+
+# string that represents the desired option
+ATTR_OPTION = "option"
```
Not being used.

lukas-hetzenecker replied:
It is used in the option service as a parameter which option to select.

balloob commented on homeassistant/components/select/__init__.py:
```diff
@@ -0,0 +1,110 @@
+
+_LOGGER = logging.getLogger(__name__)
+
+DISCOVERY_PLATFORMS = {
```
Not being used.

lukas-hetzenecker replied:
Again, this depends - my PR for HDMI adds a discovery platform.

balloob commented on homeassistant/components/select/__init__.py:
```diff
@@ -0,0 +1,110 @@
+from homeassistant.helpers.entity_component import EntityComponent
+
+DOMAIN = 'select'
+SCAN_INTERVAL = 30
```
Not being used.

maddox commented:
```yaml
allow_no_choice: yes
```

maybe: `allow_choice: no` and default it to `yes`? The negative in the property is hard to parse with your brain.

maddox commented:
I'm excited about this new component. It could be useful for things like Harmony or other multi state items :metal: :metal:.

A corresponding MQTT version would be awesome too once it's complete.

philipbl commented on homeassistant/components/select/command_select.py:
```diff
@@ -0,0 +1,70 @@
+    for dev_name, properties in selects.items():
+        name = properties.get('name', dev_name)
+        for cmd in properties.get('options', []):
+            commands.append((cmd.get('name', name), cmd.get('cmd', 'true')))
```
It might be a good idea to check that there are no duplicate names. From your logic (line 58) it looks like you select the first occurrence of that name which might not be expected by the user.

For example (it's not the best example, but hopefully it illustrates the point):

```yaml
- name: Turn on TV
- cmd: turn_on_tv_1
- name: Turn on TV
- cmd: turn_on_tv_2
```

As the user, I know the difference in the names because of their ordering. However, when I select the second option, it will execute the first command.

Since this is not a `CommandSelect` specific problem, maybe it should be implemented `select.setup` method? Any thoughts?

philipbl commented on homeassistant/components/select/__init__.py:
```diff
@@ -0,0 +1,110 @@
+            for target_select in target_selects:
+                if target_select.should_poll:
+                    target_select.update_ha_state(True)
+
```
Maybe I am missing something, but why not combine the for loops?

balloob replied:
This is the pattern we use with the other platforms as well. It prevents a throttle on update not being able to fetch all new state. However, this entity can be made into a non polling one that pushes state, removing the need for an update call.

lukas-hetzenecker replied:
This also depends if this component supports different platforms - I have an implementation of HDMI CEC for this component that uses a polling mechanism.

lukas-hetzenecker commented:
@maddox Thanks for your feedback. There was a [comprehensive discussion on gitter](https://gitter.im/balloob/home-assistant?at=568878683acb61171600a5fc) if this component should be stand-alone or not. 

As I am not familiar with MQTT, could you please elaborate on how you are planning to use it? 
I also am strongly in favor of turning this into a full-featured component with platforms - related to a switch, but for different use cases.

As you mentioned Harmony, do you mean the programmable remote control? I also got one recently and plan to integrate it with HA.

rmkraus commented on homeassistant/components/select/services.yaml:
```diff
@@ -0,0 +1,13 @@
+  fields:
+    entity_id:
+      description: Name(s) of the selectable entity
+      example: 'fan.select'
```
Should this be select.fan?

lukas-hetzenecker replied:
Yes, it should be

tomduijf commented:
Like @maddox, very exited about this indeed.
Use cases all over the place, like selecting radio channels or playlists for media_player devices, multiple light scenes for the same group of lights (although not sure how yet), various presets or modes for household equipment (like fans or heating), etc.

lukas-hetzenecker commented:
I have updated the fronted to use dropdown menus:
![s8](https://cloud.githubusercontent.com/assets/2315617/12441369/901d462e-bf46-11e5-8792-9f5e67bac14b.png)

lukas-hetzenecker commented:
How can I proceed with this component?
As can by seen by @tomduijf's reply, there are various use cases for it: One would be for use in automation scripts (like input_boolean), but another is the command_option for executing command-line scripts when selecting an entry in the options list. In future, other platforms (Media players, HDMI) would be able to use this component too.

balloob commented:
This component should focus on the pure basics: allowing the user to select an option from a list of options. It should be in the same family of components as `input_boolean`. That's why I suggested we name this `input_select` in a comment above.

Executing command line scripts can be achieved with this approach when combined with automation to hook scripts to an option being selected. But besides calling shell commands you can also call any service provided by another component, all that is already there and provided by automation/scripts.

It should not be a goal to facilitate other components so they can build their UI. Not with this component, or any component ever. Every component should always be self contained. If a media player would need a drop down menu for something, it should be built into the frontend especially for the media player with the options being contained inside the media player entity.

balloob commented:
Closing this PR because the `input_select` component has been added in #1070
